// @babylonjs/core/FlowGraph/Blocks/Execution/ControlFlow/flowGraphWhileLoopBlock.js
import { RichTypeBoolean } from "../../../flowGraphRichTypes.js";
import { RegisterClass } from "../../../../Misc/typeStore.js";
import { FlowGraphExecutionBlockWithOutSignal } from "../../../flowGraphExecutionBlockWithOutSignal.js";
/**
 * @experimental
 * A block that executes a branch while a condition is true.
 */
export class FlowGraphWhileLoopBlock extends FlowGraphExecutionBlockWithOutSignal {
    constructor(config) {
        super(config);
        this.config = config;
        this.condition = this.registerDataInput("condition", RichTypeBoolean);
        this.loopBody = this._registerSignalOutput("loopBody");
    }
    _execute(context, _callingSignal) {
        var _a;
        let conditionValue = this.condition.getValue(context);
        if (((_a = this.config) === null || _a === void 0 ? void 0 : _a.isDo) && !conditionValue) {
            this.loopBody._activateSignal(context);
        }
        while (conditionValue) {
            this.loopBody._activateSignal(context);
            conditionValue = this.condition.getValue(context);
        }
        this.out._activateSignal(context);
    }
    getClassName() {
        return FlowGraphWhileLoopBlock.ClassName;
    }
    serialize(serializationObject) {
        var _a;
        super.serialize(serializationObject);
        serializationObject.isDo = (_a = this.config) === null || _a === void 0 ? void 0 : _a.isDo;
    }
}
FlowGraphWhileLoopBlock.ClassName = "FGWhileLoopBlock";
RegisterClass(FlowGraphWhileLoopBlock.ClassName, FlowGraphWhileLoopBlock);
